hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala

```diff
@@ -0,0 +1,260 @@
+
+  if (dynamicPartitionOverwrite) {
+    // until there's explicit extensions to the PathOutputCommitProtocols
+    // to support the spark mechanism, it's left to the individual committer
+    // choice to handle partitioning.
+    throw new IOException("PathOutputCommitProtocol does not support dynamicPartitionOverwrite")
+  }
+
+  import PathOutputCommitProtocol._
+
+  /**
+   * Set up the committer.
+   * This creates it by talking directly to the Hadoop factories, instead
+   * of the V1 `mapred.FileOutputFormat` methods.
+   * @param context task attempt
+   * @return the committer to use. This will always be a subclass of
+   *         [[PathOutputCommitter]].
+   */
+  override protected def setupCommitter(
+    context: TaskAttemptContext): PathOutputCommitter = {
+
+    logInfo(s"Setting up committer for path $destination")
+    committer = PathOutputCommitterFactory.createCommitter(destPath, context)
+
+    // Special feature to force out the FileOutputCommitter, so as to guarantee
+    // that the binding is working properly.
+    val rejectFileOutput = context.getConfiguration
+      .getBoolean(REJECT_FILE_OUTPUT, REJECT_FILE_OUTPUT_DEFVAL)
+    if (rejectFileOutput && committer.isInstanceOf[FileOutputCommitter]) {
+      // the output format returned a file output format committer, which
+      // is exactly what we do not want. So switch back to the factory.
+      val factory = PathOutputCommitterFactory.getCommitterFactory(
+        destPath,
+        context.getConfiguration)
+      logInfo(s"Using committer factory $factory")
+      committer = factory.createOutputCommitter(destPath, context)
+    }
+
+    logInfo(s"Using committer ${committer.getClass}")
+    logInfo(s"Committer details: $committer")
+    if (committer.isInstanceOf[FileOutputCommitter]) {
+      require(!rejectFileOutput,
+        s"Committer created is the FileOutputCommitter $committer")
+
+      if (committer.isCommitJobRepeatable(context)) {
+        // If FileOutputCommitter says its job commit is repeatable, it means
+        // it is using the v2 algorithm, which is not safe for task commit
+        // failures. Warn
+        logWarning(s"Committer $committer may not be tolerant of task commit failures")
+      }
+    }
+    committer
+  }
+
+  /**
+   * Create a temporary file for a task.
+   *
+   * @param taskContext task context
+   * @param dir         optional subdirectory
+   * @param ext         file extension
+   * @return a path as a string
+   */
+  override def newTaskTempFile(
+    taskContext: TaskAttemptContext,
+    dir: Option[String],
+    ext: String): String = {
+
+    val workDir = committer.getWorkPath
+    val parent = dir.map(d => new Path(workDir, d)).getOrElse(workDir)
+    val file = new Path(parent, buildFilename(taskContext, ext))
+    logInfo(s"Creating task file $file for dir $dir and ext $ext")
+    file.toString
+  }
+
+  /**
+   * Absolute files are still renamed into place with a warning.
```

Comment:
Maybe explain why the warning is desirable?

Reply:
Added in the comments, not the message
1. its inefficient on any store where the copy is `O(data)`
1. If the store caches negative HEAD/GET responses briefly (i.e S3 load balancers), then if there's an attempt to copy a file almost immediately after completing the PUT for that file, then that rename may actually fail, "file not found". Which can only happen  here if the job is committed immediately after this task runs so the interval between `newTaskTempFileAbsPath()` and  `HadoopMapReduceCommitProtocol.commitJob()` is within the negative cache interval, *and* they all share the same load balancer. 

Risk #2 is those kind of problems which rarely surfaces in testing. It could be handled in `HadoopMapReduceCommitProtocol` with some retries around FileNotFoundException, if it was really found to be a problem. I've also promised the hotels.com circus-train team an option to create a file without doing that existence check...